pilot/pkg/config/kube/ior/route.go

```diff
@@ -531,60 +183,112 @@ func (r *route) createRoute(metadata config.Meta, gateway *networking.Gateway, o
 			TLS:            tlsConfig,
 			WildcardPolicy: wildcard,
 		},
-	}, metav1.CreateOptions{})
+	}
+}
+
+func (r *routeController) createRoute(
+	metadata config.Meta,
+	originalHost string,
+	tls *networking.ServerTLSSettings,
+	serviceNamespace, serviceName string,
+) (*v1.Route, error) {
+	IORLog.Debugf("creating route for hostname %s", originalHost)
+
+	nr, err := r.
+		routeClient.
+		RouteV1().
+		Routes(serviceNamespace).
+		Create(context.TODO(), buildRoute(metadata, originalHost, tls, serviceNamespace, serviceName), metav1.CreateOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating a route for the host %s from gateway: %s/%s",
+			originalHost,
+			metadata.Namespace,
+			metadata.Name)
+	}
+
+	IORLog.Infof("route %s/%s created for hostname %s from gateway %s/%s",
+		nr.ObjectMeta.Namespace, nr.ObjectMeta.Name,
+		nr.Spec.Host,
+		metadata.Namespace, metadata.Name)
+
+	return nr, nil
+}
+
+func (r *routeController) updateRoute(
+	metadata config.Meta,
+	originalHost string,
+	tls *networking.ServerTLSSettings,
+	serviceNamespace string, serviceName string,
+) (*v1.Route, error) {
+	IORLog.Debugf("updating route for hostname %s", originalHost)
+
+	nr, err := r.
+		routeClient.
+		RouteV1().
+		Routes(serviceNamespace).
+		Update(context.TODO(), buildRoute(metadata, originalHost, tls, serviceNamespace, serviceName), metav1.UpdateOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("error creating a route for the host %s (gateway: %s/%s): %s", originalHost, metadata.Namespace, metadata.Name, err)
+		return nil, errors.Wrapf(err, "error updating a route for the host %s from gateway: %s/%s",
+			originalHost,
+			metadata.Namespace,
+			metadata.Name)
 	}
 
-	IORLog.Infof("Created route %s/%s for hostname %s (gateway: %s/%s)",
+	IORLog.Infof("route %s/%s updated for hostname %s from gateway %s/%s",
 		nr.ObjectMeta.Namespace, nr.ObjectMeta.Name,
 		nr.Spec.Host,
 		metadata.Namespace, metadata.Name)
 
 	return nr, nil
 }
 
+func (r *routeController) findRoutes(metadata config.Meta) ([]*v1.Route, error) {
+	defaultLabelSet := getDefaultRouteLabelMap(metadata.Name, metadata.Namespace)
+
+	labels := labels.SelectorFromSet(defaultLabelSet)
+
+	return r.routeLister.List(labels)
+}
+
 // findService tries to find a service that matches with the given gateway selector
 // Returns the namespace and service name that is a match, or an error
-func (r *route) findService(gateway *networking.Gateway) (string, string, error) {
-	r.namespaceLock.Lock()
-	namespaces := r.namespaces
-	r.namespaceLock.Unlock()
-
+func (r *routeController) findService(gateway *networking.Gateway) (string, string, error) {
```

Comment:
please change this function to return `NamespacedName` instead of `(string, string)`.